File: backend/app/schemas/tokens.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict


class TokenUsage(BaseModel):
    user_id: int
    tier_level: int
    tier_name: str
    available_tokens: int
    total_used_tokens: int
    daily_tokens_used: int
    daily_token_limit: int
    weekly_tokens: int
    monthly_token_limit: int | None = None
    max_tokens_per_prompt: int
    refill_frequency: str
    last_refill_at: datetime | None = None
    next_refill_at: datetime | None = None
    usage_date: date

    model_config = ConfigDict(from_attributes=True)


class TokenTransactionOut(BaseModel):
    id: int
    transaction_type: str
    tokens: int
    balance_before: int
    balance_after: int
    reference_type: str | None = None
    reference_id: int | None = None
    description: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TokenTransactionList(BaseModel):
    items: list[TokenTransactionOut]
    limit: int
    offset: int

File: frontend/src/api/tokens.ts
import client from './client';

export interface TokenUsage {
  user_id: number;
  tier_level: number;
  tier_name: string;
  available_tokens: number;
  total_used_tokens: number;
  daily_tokens_used: number;
  daily_token_limit: number;
  weekly_tokens: number;
  monthly_token_limit: number | null;
  max_tokens_per_prompt: number;
  refill_frequency: string;
  last_refill_at: string | null;
  next_refill_at: string | null;
  usage_date: string;
}

export interface TokenTransaction {
  id: number;
  transaction_type: string;
  tokens: number;
  balance_before: number;
  balance_after: number;
  reference_type: string | null;
  reference_id: number | null;
  description: string | null;
  created_at: string;
}

export interface TokenTransactionList {
  items: TokenTransaction[];
  limit: number;
  offset: number;
}

export const tokensApi = {
  getUsage: async (): Promise<TokenUsage> => {
    const response = await client.get('/tokens/usage');

    return response.data;
  },

  getTransactions: async (limit = 10, offset = 0): Promise<TokenTransactionList> => {
    const response = await client.get('/tokens/transactions', {
      params: { limit, offset },
    });

    return response.data;
  },
};
